### PR TITLE
Update Pulsar Chart to add APIObject.

### PR DESCRIPTION
**Pull request by MarvinCai**
Update Pulsar Chart to add ZooKeeperCluster, BookKeeperCluster, ProxyCluster and BrokerCluster to indicate controller should manage the cluster.
And add APIObjects to indicate controller should take over existing objects.
Also add knob to disable the chart to create components that controller will create when the cluster will be managed  by controller.

### CODE REVIEW

File: charts/pulsar/templates/zookeeper/zookeeper-cluster.yaml [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+      storageClassName: {{ .Values.zookeeper.volumes.data.storageClassName }}
+      {{- end -}}
+    {{- end }}
+    reclaimPolicy: Delete
+  {{- end }}
+  config:
+    serverCnxnFactory: org.apache.zookeeper.server.NettyServerCnxnFactory
+    # copy from configmap
+    custom:
+      dataDir: /pulsar/data/zookeeper
```
Comment (fantapsody) [inline charts/pulsar/templates/zookeeper/zookeeper-cluster.yaml:183]: Maybe `dataDir` & `dataLogDir` should be configured only when `.Values.zookeeper.operator.adopt_existing` is true?

File: charts/pulsar/templates/proxy/proxy-cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      # prometheus needs to access /metrics endpoint
+      webServicePort: "{{ .Values.proxy.ports.http }}"
+      {{- if or (not .Values.tls.enabled) (not .Values.tls.proxy.enabled) }}
+      servicePort: "{{ .Values.proxy.ports.pulsar }}"
+      {{- end }}
+      {{- if and .Values.tls.enabled .Values.tls.proxy.enabled }}
+      tlsEnabledInProxy: "true"
+      servicePortTls: "{{ .Values.proxy.ports.pulsarssl }}"
+      webServicePortTls: "{{ .Values.proxy.ports.https }}"
+      tlsCertificateFilePath: "/pulsar/certs/proxy/tls.crt"
```
Comment (fantapsody) [inline charts/pulsar/templates/proxy/proxy-cluster.yaml:159]: These cert paths might only be valid when adopting an existing proxy cluster?

File: charts/pulsar/templates/proxy/proxy-cluster.yaml [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      extraOptions: []
+      memoryOptions:
+        - {{ .Values.proxy.configData.PULSAR_MEM }}
+      gcOptions:
+        - {{ .Values.proxy.configData.PULSAR_GC }}
+      gcLoggingOptions: []
+  config:
+    tls:
+      enabled: {{ and (.Values.tls.enabled) (.Values.tls.proxy.enabled) }}
+    custom:
```
Comment (fantapsody) [inline charts/pulsar/templates/proxy/proxy-cluster.yaml:119]: We may need to put custom configs in `configData` here?

File: charts/pulsar/templates/broker/broker-service.yaml [hunk 4]
```diff
@@ -27,7 +27,12 @@ metadata:
     {{- include "pulsar.standardLabels" . | nindent 4 }}
     component: {{ .Values.broker.component }}
   annotations:
+    {{- if .Values.broker.service.annotations }}
```
[the remark this answers is not in the document]
Comment (MarvinCai) [reply charts/pulsar/templates/broker/broker-service.yaml:30]: yes, I think in default value file it's just {} (including many other annotations) and causing syntax error, so I added this check.

File: charts/pulsar/templates/proxy/proxy-cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      zookeeperServers: "{{ template "pulsar.zookeeper.connect" . }}{{ .Values.metadataPrefix }}"
```
[the remark this answers is not in the document]
Comment (MarvinCai) [reply charts/pulsar/templates/proxy/proxy-cluster.yaml:120]: yeah, you're right, removed.

File: charts/pulsar/templates/bookkeeper/bookkeeper-cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        requests:
+          memory: "{{ .Values.autorecovery.resources.requests.memory }}"
+          cpu: "{{ .Values.autorecovery.resources.requests.cpu }}"
+      {{- end }}
+    conf:
+      {{- include "pulsar.bookkeeper.config.common" . | nindent 6 }}
+{{ toYaml .Values.autorecovery.configData | indent 6 }}
+{{ (.Files.Glob "conf/autorecovery/log4j2.yaml").AsConfig | indent 6 }}
+  apiObjects:
+    clientService: {}
```
[the remark this answers is not in the document]
Comment (MarvinCai) [reply charts/pulsar/templates/bookkeeper/bookkeeper-cluster.yaml:236]: actually I don't see where zk/bk client service and broker internal service is used, probably we can disable them as you suggested, will create issue to track that

File: charts/pulsar/templates/zookeeper/zookeeper-cluster.yaml (hunk 1 above)
Comment (MarvinCai) [reply charts/pulsar/templates/zookeeper/zookeeper-cluster.yaml:183]: I remember adding the check as you mentioned that to me before, somehow the change was lost, I'll add it here.

Comment (MarvinCai) [reply charts/pulsar/templates/zookeeper/zookeeper-cluster.yaml:183]: oh nvm, I only add for bk

File: charts/pulsar/templates/proxy/proxy-cluster.yaml (hunk 2 above)
Comment (MarvinCai) [reply charts/pulsar/templates/proxy/proxy-cluster.yaml:159]: that's correct, updated

File: charts/pulsar/templates/proxy/proxy-cluster.yaml (hunk 3 above)
Comment (MarvinCai) [reply charts/pulsar/templates/proxy/proxy-cluster.yaml:119]: .Values.zookeeper.configData is already

**Comment by fantapsody** [comment]
Close this since we have new plans.

**Pull request closed without merging**